gobblin-core/src/main/java/gobblin/source/extractor/extract/jdbc/TeradataExtractor.java

```diff
@@ -0,0 +1,334 @@
+      query = "SELECT COUNT(1) FROM (" + query.replace(" COUNT(1) ", " 1 at t") + ") temp";
+    }
+    commands.add(JdbcExtractor.getCommand(query, JdbcCommand.JdbcCommandType.QUERY));
+    return commands;
+  }
+
+  @Override
+  public List<Command> getDataMetadata(String schema, String entity, WorkUnit workUnit, List<Predicate> predicateList)
+      throws DataRecordException {
+    log.debug("Build query to extract data");
+    List<Command> commands = new ArrayList<Command>();
+    int fetchSize = this.workUnitState.getPropAsInt(ConfigurationKeys.SOURCE_QUERYBASED_JDBC_RESULTSET_FETCH_SIZE,
+        ConfigurationKeys.DEFAULT_SOURCE_QUERYBASED_JDBC_RESULTSET_FETCH_SIZE);
+
+    String watermarkFilter = this.concatPredicates(predicateList);
+    String query = this.getExtractSql();
+    if (isNullOrEmpty(watermarkFilter)) {
+      watermarkFilter = "1=1";
+    }
+
+    query = query.replace(ConfigurationKeys.DEFAULT_SOURCE_QUERYBASED_WATERMARK_PREDICATE_SYMBOL, watermarkFilter);
+    String sampleFilter = this.constructSampleClause();
+
+    if (!isNullOrEmpty(sampleFilter)) {
+      String columnProjection = this.getOutputColumnProjection();
+      String newColumnProjection = sampleFilter + " " + columnProjection;
+      query = query.replace(columnProjection, newColumnProjection);
+    }
+
+    commands.add(JdbcExtractor.getCommand(query, JdbcCommand.JdbcCommandType.QUERY));
+    commands.add(JdbcExtractor.getCommand(fetchSize, JdbcCommand.JdbcCommandType.FETCHSIZE));
+    return commands;    
+  }
+
+  @Override
+  public Map<String, String> getDataTypeMap() {
+    Map<String, String> dataTypeMap =
+        ImmutableMap.<String, String>builder()
+            .put("byteint", "int")
+            .put("smallint", "int")
+            .put("integer", "int")
+            .put("bigint", "long")
+            .put("float", "float")
+            .put("decimal", "double")
+            .put("char", "string")
+            .put("varchar", "string")
+            .put("byte", "bytes")
+            .put("varbyte", "bytes")
+            .put("date", "date")
+            .put("time", "time")
+            .put("timestamp", "timestamp")
+            .put("clob", "string")
+            .put("blob", "string")
+            .put("structured udt", "array")
+            .build();
+    return dataTypeMap;
+  }
+
+  @Override
+  public Iterator<JsonElement> getRecordSetFromSourceApi(String schema, String entity, WorkUnit workUnit,
+      List<Predicate> predicateList) throws IOException {
+    return null;
+  }
+
+  @Override
+  public String getConnectionUrl() {
+    String host = this.workUnit.getProp(ConfigurationKeys.SOURCE_CONN_HOST_NAME);
+    String port = this.workUnit.getProp(ConfigurationKeys.SOURCE_CONN_PORT,"1025");
+    String database = this.workUnit.getProp(ConfigurationKeys.SOURCE_QUERYBASED_SCHEMA);
+    String url = "jdbc:teradata://" + host.trim() + "/DATABASE=" + database.trim() + ",DBS_PORT=" + port.trim() ;
```

Comment:
When using the Teradata JDBC driver the `TYPE=FASTEXPORT` parameter could be added to the connection string. Optionally a default charset as well (e.g: `CHARSET=utf16`).